### ISSUE TITLE
support "classnames" syntax in :class

### ISSUE THREAD

**Issue by luwenjin** [vuejs/vue#2530]
From knockout to angular to react, now i'm new to Vue. Really love this library, feels like "angular+react for humans".

:class now support array **OR** object, but sometimes it's still a little tedious
if :class support classnames(https://github.com/JedWatson/classnames) like syntax will be better.

example:
from
`:class="[type, checked?'checked':'', readOnly?'read-only':'', disabled?'disabled':'', fitted?'fitted':''"`
into
`:class="type, {checked, disabled, fitted}, readOnly&&'read-only'"`

Sure, I can always use a computed method, but by doing this class names are defined in one more place(props+template -> props+template+computed), which harms readability.

**Comment by rpkilby** [comment]
This is not a real answer, but for readability in the short term you could break the parameter across multiple lines. eg,

```html
<div :class="{
 'type': true,
 'checked': checked,
 'read-only': readOnly,
 'disabled': disabled,
 'fitted': fitted}">
</div>
```

**Comment by fullfs** [comment]
@luwenjin Hello! I'm also having problems with multiple conditioned classnames like in your example. But please note you're proposing to bring new directive syntax for a sole case. It does not look to like a wise decision to me, because @yyx990803 tried to make directives' syntax as compact as possible and make it look like common javascript with simple expressions, arrays and objects.

@rpkilby 's version looks better to me :)

**Comment by simplesmiler** [comment]
@luwenjin `{ checked, disabled, fitted }` is ES6 and will only. Other than that you can make a custom method `cn: (classes) => classNames(classes)` and use that:
```
:class="cn(type, { checked, disabled, fitted, 'ready-only': readOnly })"
```
Or, with support for ES5 browsers:
```
:class="cn(type, { checked: checked, disabled: disabled, fitted: fitted, 'ready-only': readOnly })"
```

**Comment by luwenjin** [comment]
@rpkilby in my example, the read-only class is a special case showing false value will be ignored. In common case, it will be like:
`:class="type, {checked, disabled, fitted, readonly}'"`
it's ES5 object literal vs ES6, if you can use ES6 style, you will almost never use the old style.

@fullfs I like the compact feeling of Vue. But just like v-for, IMHO, if one directive is used so frequently, it's worth breaking the rule in order to make the user wring code faster/clearer.
And in my proposal, it's still common javascript in some way, it's the spread arguments(like ...args in function(...args){}). 
Even if sticking to the list/object type, 
`:class="[type, {checked, disabled, fitted}]'"`
it is still looking better

I admit the ES6 style object literal syntax is the hard part, and without supporting it, my proposal is almost meaningless. Hard to find an elegant way to support it in ES5 runtime

**Comment by rhyzx** [comment]
@luwenjin there is a tricky way to this, `<div :class.1="{checked: true}" :class.2="'typeName'"></div>`, notice these  `.1` and `.2` modifiers, they must be unique, else vue will ignore duplicated directives.

**Comment by rhyzx** [comment]
@yyx990803 maybe we can allow duplicated `class` directives?

**Comment by simplesmiler** [comment]
@luwenjin the problem is that while you can transpile js from ES6 to ES5, but you can not (yet) transpile expressions inside the template.

@rhyzx nice trick. Duplicates are not possible because browser parser will strip the second one out.

Regarding the issue, I think it's possible to improve `:class` syntax by supporting objects inside of a list (currently Vue only supports strings inside of a list). This will cover a lot of use cases.

**Comment by fullfs** [comment]
@simplesmiler Agree with last one. Also like @rhyzx 's trick.

P.S. It's also possible to use some custom... whatever: https://jsfiddle.net/w6608xw0/1/
("improved" @simplesmiler 's example)

**Comment by luwenjin** [comment]
@rhyzx  wow, i never thought about it, nice trick

@simplesmiler 
(about transpile) yes, exactly
I agree, objects inside list is nice to have

I think we can divide this proposal into two parts:
1. is the new syntax best for the users ? 
2. should/can Vue support it ?

IMO, 
1. yes
2. yes, but it has costs and needs to be consider

**Comment by Nirazul** [comment]
When a tag like `:class` grows to those dimensions so I have to start thinking about how to write it more elegantly, I'd prefer to just throw out the whole thing and handle it in javascript: `:class="gimmeMyClasses()"`

There'll be an overhead when I want to use it in different locations with different configurations, but in my opinion it's preferable to introducing new syntax sugar.

**Comment by yyx990803** [comment]
Here's what I think:

1. `:class` bindings should still be valid JavaScript expression.

2. We should just support objects inside arrays, which shouldn't be a lot of work:

  ``` html
  <div :class="[a, { b: b, c: c }]">
  ```

**Comment by FadyMak** [comment]
Working on it now :)

**Comment by luwenjin** [comment]
@yyx990803  nice!

**Comment by yyx990803** [comment]
Landed in 1.0.19.
